MariaDB/server
# WIP - MDEV-33408 Initial support for vector DELETE and UPDATE

- [x] *The Jira issue number for this PR is: MDEV-33408*


## Description
When the source row is deleted, mark the corresponding node in HNSW index by setting `tref` to null. An index is added for the `tref` in secondary table for faster searching of the to-be-marked nodes.

The nodes marked as deleted will still be used for search, but will not be included in the final query results.

As skipping deleted nodes and not adding deleted nodes for new-inserted nodes' neighbor list could impact the performance, we now only skip these nodes in search results.

TODO:
- for some reason the bitmap is not set for hlindex during the delete so I had to temporarily comment out [this line](https://github.com/MariaDB/server/blob/bb-11.6-MDEV-32887-vector/sql/sql_base.cc#L9874)

## Release Notes
N/A - Vector support development

## How can this PR be tested?

MTR test has been updated to verify the DELETE and UPDATE.

Benchmarking test had been done with `DATASET=random-s-100-euclidean` and `ef_construction_multiplier = 1` ( have to set this to 1 because default 4 always returns recall of 1 and hard to compare ).

#### Before
```
Insert time:  195.7592477798462
Recall:       0.872
QPS:          701.594
```

#### After
first run
```
Insert time:  218.7464780807495
Recall:       0.885
QPS:          722.408
```
second run
```
Insert time:  172.49292421340942
Recall:       0.892
QPS:          778.593
```



If the changes are not amenable to automated testing, please explain why not and carefully describe how to test manually.


## Basing the PR against the correct MariaDB version


## PR quality check

## Copyright
All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.
